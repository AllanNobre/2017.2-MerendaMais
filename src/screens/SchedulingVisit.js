import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text, View, Picker, TouchableOpacity, Alert, ScrollView, Dimensions } from 'react-native';
import PopupDialog, {
  DialogTitle,
  DialogButton,
} from 'react-native-popup-dialog';
import { Actions } from 'react-native-router-flux';
import { Ionicons } from '@expo/vector-icons';
import DatePicker from 'react-native-datepicker';
import SchoolData from '../components/SchoolData';
import Button from '../components/Button';
import store from '../Reducers/store';

const { width } = Dimensions.get('window');


const styles = StyleSheet.create({

  principal: {
    flex: 1,
  },
  textLogo: {
    // Font size 30 looks nice on 360 width phone.
    // (x * widthYourPhone = fontSize) where x is the proportion used in fontSize above.
    fontSize: width * 0.08,
    color: 'white',
    fontWeight: 'bold',
    marginTop: 10,
    marginLeft: 60,
  },
  wrapper: {
    height: 100,
    flexDirection: 'row',
    paddingTop: 10,
    paddingBottom: 10,
    backgroundColor: '#FF9500',
    borderBottomColor: 'black',
    borderBottomWidth: 1,
    alignItems: 'center',
  },
  schedullingButton: {
    paddingVertical: 20,
    borderWidth: 1,
    borderRadius: 7,
    marginHorizontal: 20,
    marginTop: 10,
    marginBottom: 10,
    backgroundColor: '#FF9500',
    justifyContent: 'flex-end',
  },

  disabledSchedullingButton: {
    paddingVertical: 20,
    borderWidth: 1,
    borderRadius: 7,
    marginHorizontal: 20,
    marginTop: 10,
    marginBottom: 10,
    backgroundColor: '#DEDEDE',
    justifyContent: 'flex-end',
  },

  button: {
    paddingVertical: 15,
    borderWidth: 1,
    borderRadius: 7,
    marginHorizontal: 20,
    marginTop: 10,
    marginBottom: 10,
    backgroundColor: '#FF9500',
    justifyContent: 'flex-end',
  },

  buttonText: {
    textAlign: 'center',
    color: '#FFF',
  },

  Container: {
    flex: 1,
    marginTop: 20,
  },
  icon_header: {
    marginLeft: 20,
  },
  Picker: {
    marginHorizontal: 15,
    width: '95%',
  },

  popUp: {
    flex: 1,
    alignItems: 'center',
    margin: 20,
  },

  popUpText: {
    fontSize: 15,
    textAlign: 'justify',
    lineHeight: 20,
  },

  popUpTitle: {
    fontSize: 30,
  },

  icon: {
    width: 20,
    height: 20,
    margin: 5,
  },

});

export default class SchedulingVisit extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      appToken: this.props.counselor.token,
      nuvemCode: this.props.counselor.nuvemCode,
      visit: {
        codSchool: 0,
        date: '',
        time: '',
        invitedAgent: false,
        agentEmail: '',
      },
    };
  }

  componentWillReceiveProps(newProps) {
    const newVisit = {
      codSchool: newProps.school.schoolCode,
      date: this.state.visit.date,
      time: this.state.visit.time,
      invitedAgent: this.state.visit.invitedAgent,
      agentEmail: this.state.visit.agentEmail,
    };

    this.setState({ visit: newVisit });
  }

  invitingAgent() {
    this.setState({ visit: { ...this.state.visit, invitedAgent: true } });
    this.popupDialog.dismiss();
  }

  notInvitingAgent() {
    this.setState({ visit: { ...this.state.visit, invitedAgent: false } });
    this.popupDialog.dismiss();
  }

  buttonActivation() {
    if (this.state.visit.agentEmail > '') {
      return (
        <DialogButton
          enabled
          key="invitingButton"
          text="Convidar"
          onPress={() => { this.invitingAgent(); }}
        />
      );
    }
    return (
      <DialogButton
        enabled
        key="notInvitingButton"
        text="Cancelar"
        onPress={() => { this.notInvitingAgent(); }}
      />
    );
  }

  testa() {
    console.log('DEFINITIVO');
    console.log(this.state.visit);
  }
  render() {
    const newStateDoc = store.getState();
    console.log(newStateDoc.schedule);

    return (
      <View style={styles.principal}>
        <View style={styles.wrapper}>
          <TouchableOpacity onPress={() => Actions.mainScreen()} >
            <Ionicons
              name="ios-arrow-back-outline"
              style={styles.icon_header}
              size={45}
              color="black"
            />
          </TouchableOpacity>
          <Text style={styles.textLogo}>Agendar Visita</Text>
        </View>

        <PopupDialog
          ref={(popupDialog) => { this.popupDialog = popupDialog; }}
          dialogTitle={<DialogTitle
            title="Convidar um Agente?"
            titleStyle={styles.popUpTitle}
          />}
        >
          <View style={styles.popUp}>
            <Text style={styles.popUpText}>Escolha algum dos agentes abaixo para poder convidá-lo.
            Para isso, é necessário possuir um aplicativo de email instalado
            no seu celular. Caso não possua ou não deseje convidar um agente,
            não selecione nenhum agente e clique em cancelar.</Text>

            <Picker
              style={styles.Picker}
              selectedValue={this.state.visit.agentEmail}
              onValueChange={
                value => this.setState({ visit: { ...this.state.visit, agentEmail: value } })}
            >
              <Picker.Item value="" label="Escolha o agente" color="#95a5a6" />
              <Picker.Item value={'outroemail@email.com'} label={'Agente Sanitário'} />
              <Picker.Item value={'email@email.com'} label={'Poder Executivo'} />
            </Picker>
            {this.buttonActivation()}
          </View>
        </PopupDialog>

        <ScrollView>
          <View style={styles.Container}>
            <View>
              <TouchableOpacity
                key="searchSchoolButton"
                style={styles.button}
                onPress={() => Actions.searchSchool()}
              >
                <Text style={styles.buttonText}>Pesquisar escola</Text>
              </TouchableOpacity>
            </View>
          </View>

          {this.props.school.schoolSelected && (
            <SchoolData {...this.props.school} />
          )}

          <View style={styles.Container}>
            <DatePicker
              style={styles.Picker}
              placeholder="Data"
              date={this.state.visit.date}
              mode="date"
              format="DD-MM-YYYY"
              confirmBtnText="Confirmar"
              cancelBtnText="Cancelar"
              customStyles={{
                dateInput: {
                  borderRadius: 7,
                },
              }}
              onDateChange={date => this.setState({ visit: { ...this.state.visit, date } })}
            />
          </View>

          <View style={styles.Container}>
            <DatePicker
              style={styles.Picker}
              placeholder="Horário"
              date={this.state.visit.time}
              mode="time"
              confirmBtnText="Confirmar"
              cancelBtnText="Cancelar"
              customStyles={{
                dateInput: {
                  borderRadius: 7,
                },
              }}
              onDateChange={time => this.setState({ visit: { ...this.state.visit, time } })}
            />
          </View>

          <View style={styles.Container}>
            <View>
              <TouchableOpacity
                key="searchCounselorButton"
                style={styles.button}
                onPress={() => Alert.alert('Pesquisando')}
              >
                <Text style={styles.buttonText}>Pesquisar Conselheiro</Text>
              </TouchableOpacity>
            </View>
          </View>

          <View style={styles.Container}>
            <View>
              <TouchableOpacity
                key="searchAgentButton"
                style={styles.button}
                onPress={() => this.popupDialog.show()}
              >
                <Text style={styles.buttonText}>Convidar Agente</Text>
              </TouchableOpacity>
            </View>
          </View>
          <View style={styles.Container}>
            {this.props.school.schoolSelected && (
              <Button
                enabled
                key="scheduleButton"
                text="Agendar"
                onPress={() => {
                  Alert.alert(
                    'Agendamento Realizado',
                    'O agendamento foi realizado com sucesso! Caso tenha convidado um agente, seu aplicativo de email abrirá.',
                    [
                      { text: 'Ok', onPress: () => this.props.asyncSchedulingVisit(this.state), style: 'cancel' },
                    ],
                    { cancelable: false });
                }}
              />
            )}
            <Button
              enabled={false}
              text="Agendar"
              key="scheduleButton"
              onPress={() => ({})}
              disabled
            />

          </View>
        </ScrollView>
      </View>
    );
  }
}

const { shape, string, number } = PropTypes;

SchedulingVisit.propTypes = {
  asyncSchedulingVisit: PropTypes.func.isRequired,
  counselor: shape({
    token: string.isRequired,
    nuvemCode: number.isRequired,
  }).isRequired,
  school: shape({
    schoolCode: number.isRequired,
    schoolName: string.isRequired,
    schoolPhone: string.isRequired,
    schoolEmail: string.isRequired,
  }).isRequired,
};
