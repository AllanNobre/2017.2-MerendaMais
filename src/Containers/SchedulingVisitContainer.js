import { connect } from 'react-redux';
import SchedulingVisit from '../screens/SchedulingVisit';
import asyncSchedulingVisit from '../actions/schedulingVisitActions';
import { asyncGetCounselorFromGroup,
  setNewLists } from '../actions/listActions';

const mapStateToProps = state => ({
  listOfCounselorsInAGroup: state.list.listOfCounselorsInAGroup,
  listOfInviteesWithCounselorInformations: state.list.listOfInviteesWithCounselorInformations,
  listOfInvitees: state.list.listOfInvitees,
  counselor: state.counselor,
  school: state.school,
});

const mapDispatchToProps = dispatch => ({
  asyncGetCounselorFromGroup: (CAE, CPF) => dispatch(asyncGetCounselorFromGroup(CAE, CPF)),

  asyncSchedulingVisit: visitData => dispatch(asyncSchedulingVisit(visitData)),

  setNewLists: newListWithInformations =>
    dispatch(setNewLists(newListWithInformations)),
});

const SchedulingVisitContainer = connect(mapStateToProps, mapDispatchToProps)(SchedulingVisit);

export default SchedulingVisitContainer;
