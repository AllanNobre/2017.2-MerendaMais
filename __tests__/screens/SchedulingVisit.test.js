import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import SchedulingVisit from '../../src/screens/SchedulingVisit';
import SchedulingVisitContainer from '../../src/Containers/SchedulingVisitContainer';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();

const initialState = {
  counselor: {
    nuvemCode: 1,
    token: 'tokenGenerico',
  },
  visit: {
    codSchool: 32,
    date: '31-12-2017',
    time: '18:30',
  },
  school: {
    schoolSelected: true,
  },
};

const store = mockStore(initialState);

jest.mock('react-native-router-flux');

describe('Testing SchedulingVisit', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <SchedulingVisitContainer />,
      { context: { store } },
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Testing SchedulingVisit buttons', () => {
  it('Test if search school button is rendered', () => {
    const wrapper = shallow(<SchedulingVisit {...initialState} />);
    const button = wrapper.findWhere(c => c.key() === 'searchSchoolButton');
    expect(button.length).toEqual(1);
    button.simulate('press');
  });
  it('Test if search counselor button is rendered', () => {
    const wrapper = shallow(<SchedulingVisit {...initialState} />);
    const button = wrapper.findWhere(c => c.key() === 'searchCounselorButton');
    expect(button.length).toEqual(1);
    button.simulate('press');
  });
  it('Test if search agent button is rendered', () => {
    const wrapper = shallow(<SchedulingVisit {...initialState} />);
    const button = wrapper.findWhere(c => c.key() === 'searchAgentButton');
    expect(button.length).toEqual(1);
    button.simulate('press');
  });

  it('Test if schedule button is rendered', () => {
    const asyncSchedulingVisit = (state) => {
      expect(state.nuvemCode).toEqual(1);
      expect(state.token).toEqual('tokenGenerico');
      expect(state.visit.codSchool).toEqual(32);
      expect(state.visit.date).toEqual('31-12-2017');
      expect(state.visit.time).toEqual('18:30');
    };

    const wrapper = shallow(<SchedulingVisit
      {...initialState}
    />);

    wrapper.setState({
      nuvemCode: 1,
      token: 'tokenGenerico',
      visit: {
        codSchool: 32,
        date: '31-12-2017',
        time: '18:30',
      },
    });

    const button = wrapper.findWhere(c => c.key() === 'scheduleButton');
    expect(button.length).toEqual(2);
  });
});

describe('Testing SchedulingVisit DatePickers', () => {
  const wrapper = shallow(<SchedulingVisit {...initialState} />);

  it('Test if date changes when DatePicker changes', () => {
    const datePickerComponent = wrapper.find('DatePicker').at(0);
    datePickerComponent.simulate('dateChange', '31-10-2017');
    expect(wrapper.state().visit.date).toEqual('31-10-2017');
  });
  it('Test if time changes when DatePicker changes', () => {
    const datePickerComponent = wrapper.find('DatePicker').at(1);
    datePickerComponent.simulate('dateChange', '19:50');
    expect(wrapper.state().visit.time).toEqual('19:50');
  });
});
