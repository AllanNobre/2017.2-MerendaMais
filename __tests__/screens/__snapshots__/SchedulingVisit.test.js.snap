// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SchedulingVisit renders as expected 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF9500",
        "borderBottomColor": "black",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "height": 100,
        "paddingBottom": 10,
        "paddingTop": 10,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Icon
        allowFontScaling={false}
        color="black"
        name="ios-arrow-back-outline"
        size={45}
        style={
          Object {
            "marginLeft": 20,
          }
        }
      />
    </TouchableOpacity>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 60,
          "fontWeight": "bold",
          "marginLeft": 60,
          "marginTop": 10,
        }
      }
    >
      Agendar Visita
    </Text>
  </View>
  <PopupDialog
    dialogTitle={
      <DialogTitle
        haveTitleBar={true}
        title="Convidar um Agente?"
        titleAlign="center"
        titleStyle={
          Object {
            "fontSize": 30,
          }
        }
      />
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "margin": 20,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 15,
            "lineHeight": 20,
            "textAlign": "justify",
          }
        }
      >
        Escolha algum dos agentes abaixo para poder convidá-lo. Para isso, é necessário possuir um aplicativo de email instalado no seu celular. Caso não possua ou não deseje convidar um agente, não selecione nenhum agente e clique em cancelar.
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue=""
        style={
          Object {
            "marginHorizontal": 15,
            "width": "95%",
          }
        }
      >
        <PickerItem
          color="#95a5a6"
          label="Escolha o agente"
          value=""
        />
        <PickerItem
          label="Agente Sanitário"
          value="outroemail@email.com"
        />
        <PickerItem
          label="Poder Executivo"
          value="email@email.com"
        />
      </Picker>
      <DialogButton
        align="center"
        disabled={false}
        enabled={true}
        key="notInvitingButton"
        onPress={[Function]}
        text="Cancelar"
      />
    </View>
  </PopupDialog>
  <ScrollViewMock>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <View>
        <TouchableOpacity
          activeOpacity={0.2}
          key="searchSchoolButton"
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#FF9500",
              "borderRadius": 7,
              "borderWidth": 1,
              "justifyContent": "flex-end",
              "marginBottom": 10,
              "marginHorizontal": 20,
              "marginTop": 10,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#FFF",
                "textAlign": "center",
              }
            }
          >
            Pesquisar escola
          </Text>
        </TouchableOpacity>
      </View>
    </View>
    <SchoolData
      schoolSelected={true}
    />
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <DatePicker
        TouchableComponent={[Function]}
        androidMode="default"
        cancelBtnText="Cancelar"
        confirmBtnText="Confirmar"
        customStyles={
          Object {
            "dateInput": Object {
              "borderRadius": 7,
            },
          }
        }
        date=""
        disabled={false}
        duration={300}
        format="DD-MM-YYYY"
        height={259}
        hideText={false}
        iconSource={1}
        modalOnResponderTerminationRequest={[Function]}
        mode="date"
        onDateChange={[Function]}
        placeholder="Data"
        showIcon={true}
        style={
          Object {
            "marginHorizontal": 15,
            "width": "95%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <DatePicker
        TouchableComponent={[Function]}
        androidMode="default"
        cancelBtnText="Cancelar"
        confirmBtnText="Confirmar"
        customStyles={
          Object {
            "dateInput": Object {
              "borderRadius": 7,
            },
          }
        }
        date=""
        disabled={false}
        duration={300}
        height={259}
        hideText={false}
        iconSource={1}
        modalOnResponderTerminationRequest={[Function]}
        mode="time"
        onDateChange={[Function]}
        placeholder="Horário"
        showIcon={true}
        style={
          Object {
            "marginHorizontal": 15,
            "width": "95%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <View>
        <TouchableOpacity
          activeOpacity={0.2}
          key="searchCounselorButton"
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#FF9500",
              "borderRadius": 7,
              "borderWidth": 1,
              "justifyContent": "flex-end",
              "marginBottom": 10,
              "marginHorizontal": 20,
              "marginTop": 10,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#FFF",
                "textAlign": "center",
              }
            }
          >
            Pesquisar Conselheiro
          </Text>
        </TouchableOpacity>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <View>
        <TouchableOpacity
          activeOpacity={0.2}
          key="searchAgentButton"
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#FF9500",
              "borderRadius": 7,
              "borderWidth": 1,
              "justifyContent": "flex-end",
              "marginBottom": 10,
              "marginHorizontal": 20,
              "marginTop": 10,
              "paddingVertical": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#FFF",
                "textAlign": "center",
              }
            }
          >
            Convidar Agente
          </Text>
        </TouchableOpacity>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <Button
        enabled={true}
        key="scheduleButton"
        onPress={[Function]}
        text="Agendar"
      />
      <Button
        disabled={true}
        enabled={false}
        key="scheduleButton"
        onPress={[Function]}
        text="Agendar"
      />
    </View>
  </ScrollViewMock>
</View>
`;
